### ISSUE TITLE
possible OOS with double random seed requests

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#3925]
a few days a go i observed a mp game that gave OOS and the replay contained the following sequence
<details>
 <summary>Replay Code</summary>
```
	[command]
		from_side=4
		sent=yes
		[recruit]
			type="Gryphon Rider"
			x=20
			y=5
			[from]
				x=21
				y=6
			[/from]
		[/recruit]
		[checkup]
			[result]
				checksum="43.-Em6fQO<J,dopZ5,D65305j8mk61gh7fd>93,a6nkmg,,b-eLmo3.<ko.X2xOw,9hi7kpeY<Voj7h-8<7,m83Ca+9<dd<,ijbjq<2fQ9MCDx<nuTF+Sug5alqPwTR"
			[/result]
			[result]
				next_unit_id=145
				random_calls=22
			[/result]
		[/checkup]
	[/command]
	[command]
		dependent=yes
		from_side="server"
		sent=yes
		[random_seed]
			new_seed="5096b4ff"
		[/random_seed]
	[/command]
	[command]
		sent=yes
		undo=no
		[speak]
			id="Aquamarin"
			message="undead my side still i have known it"
			side=1
			time=1549732743
		[/speak]
	[/command]
	[command]
		sent=yes
		undo=no
		[speak]
			id="Aquamarin"
			message="undead my side still i have known it"
			side=1
			time=1549732743
		[/speak]
	[/command]
	[command]
		dependent=yes
		from_side="server"
		sent=yes
		[random_seed]
			new_seed="6de69e5f"
		[/random_seed]
	[/command]
	[command]
		sent=yes
		undo=no
		[speak]
			id="server"
			message="Ado64 reports out of sync errors."
		[/speak]
	[/command]
	[command]
		sent=yes
		undo=no
		[speak]
			id="server"
			message="flynnk reports out of sync errors."
		[/speak]
	[/command]
	[command]
		from_side=4
		sent=yes
		[recruit]
			type="Thief"
			x=21
			y=7
			[from]
				x=21
				y=6
			[/from]
		[/recruit]
		[checkup]
			[result]
				checksum="3l27Hac<VYYqc8>1Uba5-fxt4,9f0-7s,2J6a0<Aa<,1dF53877eer42iY1dlWfn>bhX5+m3.S6f,<3g,wcEsLqhThmEu<bDS<Ib55Q>,.6tmRsA.zbLBHJULMiyQCKP"
			[/result]
			[result]
				next_unit_id=146
				random_calls=23
			[/result]
		[/checkup]
	[/command]

```
</details>

The important thing here is that there were two [random_seed] after the first [recruit] command whihc shouldn't really be possible. Of course it it not 100% excuded that this is caused by people that changed data files but i cuccrently don't know of any way that would result this type of OOS. It should also be noticed that this happend during an ai turn so undoing and dsu which were common sources of OOS in he past were probably not involved here.

**Comment by gfgtdf** [comment]
Okay lets think about it some more:

The server sended two different 'random_seed' during one action, why could that be? The only cause could be that is recives two requests with different ids during the execution of that  actions, what could  be the reason?
1) Another client sended the 'request random seed' message after it received the servers original random seed this is highly unlikely
2) Another client sended the 'request random seed' message before  he recived the original random_seed message, this however imples that his request_id at that point was above the request id of the server after he sended the first message (othwesie the server woudl have discarded this request) , so this in particular means that the server request_id values on the two client were different before.


So we for now assume possiblity (2) from above, the next question  is how could  we get into such a  situation, or rather :how could  we get into such a  situation without getting an OOS mesage before?

Clearly the request_id counter can become desynced with due to an earlier OOS one client woudl assume that an event requires a random choice but another doesn't. But that situation should also have given the players an  OOS  dialog earlier.

**Comment by soliton-** [comment]
Is Aquamarine’s duplicate chat message a copy&paste error or real? Perhaps a hint that some code ran twice for some reason.

**Comment by gfgtdf** [comment]
> Is Aquamarine’s duplicate chat message a copy&paste error or real?

i think its real, i see no reason how i could ahve made a vcopy and past error there. but im not 100% sure, i also don't think i could easily find that replay.

**Comment by soliton-** [comment]
It's this replay: http://replays.wesnoth.org/1.14/2019/02/09/WC_II_3_+%23textdomain_wesnoth-World_Conquest_II_players%5ep_+_-__+_Scenario_+_2_-__+_Provinces_Turn_2_(8665).bz2

**Comment by gfgtdf** [comment]
> It's this replay: http://replays.wesnoth.org/1.14/2019/02/09/WC_II_3_+%23textdomain_wesnoth-World_Conquest_II_players%5ep_+_-__+_Scenario_+_2_-__+_Provinces_Turn_2_(8665).bz2

oh wow thx :), okay this answers your question the message is so there twice in this file. so it was not a copypaste error

**Comment by gfgtdf** [comment]
somethign i have noticed in the server replay: it lacks the '... reports out fo sync errors' messages, this is pbobably an oversight in the servr code but woudl be nice to be fixed since it could help detecting oos.

**Comment by soliton-** [comment]
Those messages are not saved in the replay indeed. The termination reason of a game is logged though and it is set to "out of sync" in such cases. For some reason this game ended with reason 'game over' though so at least one client must have sent that instead as the last termination reason.

Saving the server messages in the replay as well is not a bad idea.

**Comment by gfgtdf** [comment]
> The termination reason of a game is logged though and it is set to "out of sync" in such cases

hmm yes but for debugging it'd be nice to know when was the first time when the OOS was noticed.

> Saving the server messages in the replay as well is not a bad idea.

Yes we should  do that, appearently game::send_and_record_server_message already exists but soem cases call game::send_server_message instead, shodul we jsut replace all occurances with the second one  with the first one?

**Comment by soliton-** [comment]
> shodul we jsut replace all occurances with the second one with the first one?

No. I already looked and besides the OOS message nothing else stuck out as making much sense to save in the replay since they're responses to client commands lick kick/ban/mute etc.

You could port the OOS message commit to 1.14 if you like.

**Comment by gfgtdf** [comment]
I don't think we can find the underlying cause of this, what we can do however is to store re request_id in the response so all clients can check whether it matches their assumed request Id, and give an error Message earlier when something goes wrong, after that I'd probably close this issue
